cdap-data-fabric/src/main/java/co/cask/cdap/data2/datafabric/dataset/type/LocalDatasetTypeClassLoaderFactory.java

```diff
@@ -16,24 +16,57 @@
 
 package co.cask.cdap.data2.datafabric.dataset.type;
 
+import co.cask.cdap.common.lang.jar.BundleJarUtil;
+import co.cask.cdap.common.utils.DirUtils;
 import co.cask.cdap.proto.DatasetModuleMeta;
+import co.cask.common.lang.DirectoryClassLoader;
+import com.google.common.hash.Hashing;
+import com.google.common.io.Files;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
 
+import java.io.File;
 import java.io.IOException;
-import java.net.URL;
-import java.net.URLClassLoader;
+import java.net.URI;
 
 /**
- * Creates a {@link ClassLoader} for a {@link DatasetModuleMeta} without unpacking the moduleMeta jar,
- * since the jar file is already present in the local filesystem.
+ * Creates a {@link ClassLoader} for a {@link DatasetModuleMeta} in local mode. Unpacks the jar if it is not unpacked
+ * already, and uses a {@link DirectoryClassLoader} on the unpacked jar.
  */
 public class LocalDatasetTypeClassLoaderFactory implements DatasetTypeClassLoaderFactory {
+  private static final Logger LOG = LoggerFactory.getLogger(LocalDatasetTypeClassLoaderFactory.class);
 
   @Override
   public ClassLoader create(DatasetModuleMeta moduleMeta, ClassLoader parentClassLoader) throws IOException {
-    if (moduleMeta.getJarLocation() == null) {
+    URI jarLocation = moduleMeta.getJarLocation();
+    if (jarLocation == null) {
       return parentClassLoader;
     }
 
-    return new URLClassLoader(new URL[]{moduleMeta.getJarLocation().toURL()}, parentClassLoader);
+    File jarFile = new File(jarLocation);
+    String dirName = String.format("%s.%s", moduleMeta.getName(), Files.hash(jarFile, Hashing.md5()).toString());
+    File expandDir = new File(jarFile.getParent(), dirName);
```

Comment:
deleting this correctly will require some refactoring up the stack to make sure things are closed properly. Opened https://issues.cask.co/browse/CDAP-2564 to do this.